src/Xamarin.Android.Build.Tasks/Utilities/XAAssemblyResolver.cs

```diff
@@ -0,0 +1,333 @@
+	}
+
+	public AssemblyDefinition Resolve (AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		return Resolve (AndroidTargetArch.None, name, parameters);
+	}
+
+	public AssemblyDefinition Resolve (AndroidTargetArch arch, AssemblyNameReference name, ReaderParameters? parameters = null)
+	{
+		string shortName = name.Name;
+		if (cache.TryGetValue (shortName, out CacheEntry? entry)) {
+			return SelectAssembly (arch, name.FullName, entry, loading: false);
+		}
+
+		if (arch == AndroidTargetArch.None) {
+			return FindAndLoadFromDirectories (arch, FrameworkSearchDirectories, name, parameters);
+		}
+
+		if (!AbiSearchDirectories.TryGetValue (arch, out ICollection<string>? directories) || directories == null) {
+			throw CreateLoadException (name);
+		}
+
+		return FindAndLoadFromDirectories (arch, directories, name, parameters);
+	}
+
+	AssemblyDefinition FindAndLoadFromDirectories (AndroidTargetArch arch, ICollection<string> directories, AssemblyNameReference name, ReaderParameters? parameters)
+	{
+		string? assemblyFile;
+		AssemblyDefinition? candidate = null;
+		foreach (string dir in directories) {
+			if ((assemblyFile = SearchDirectory (name.Name, dir)) != null) {
+				AssemblyDefinition? loaded = Load (arch, assemblyFile);
+				if (Array.Equals (loaded?.Name.MetadataToken, name.MetadataToken)) {
+					return loaded;
+				}
+				candidate = candidate ?? loaded;
+			}
+		}
+
+		if (candidate != null)
+			return candidate;
+
+		throw CreateLoadException (name);
+	}
+
+	static FileNotFoundException CreateLoadException (AssemblyNameReference name)
+	{
+		return new FileNotFoundException ($"Could not load assembly '{name}'.");
+	}
+
+	static string? SearchDirectory (string name, string directory)
+	{
+		if (Path.IsPathRooted (name) && File.Exists (name)) {
+			return name;
+		}
+
+		var file = Path.Combine (directory, $"{name}.dll");
+		if (File.Exists (file)) {
+			return file;
+		}
+
+		return null;
+	}
+
+	public virtual AssemblyDefinition? Load (AndroidTargetArch arch, string filePath, bool forceLoad = false)
+	{
+		string name = Path.GetFileNameWithoutExtension (filePath);
+		AssemblyDefinition? assembly;
+		if (!forceLoad && cache.TryGetValue (name, out CacheEntry? entry)) {
+			assembly = SelectAssembly (arch, name, entry, loading: true);
+			if (assembly != null) {
+				return assembly;
+			}
+		}
+
+		try {
+			assembly = ReadAssembly (filePath);
+		} catch (Exception e) when (e is FileNotFoundException || e is DirectoryNotFoundException) {
+			// These are ok, we can return null
+			return null;
+		}
+
+		if (!cache.TryGetValue (name, out entry)) {
+			entry = new CacheEntry (log, filePath, assembly, arch);
+			cache.Add (name, entry);
+		} else {
+			entry.Add (arch, assembly);
+		}
+
+		return assembly;
+	}
+
+	AssemblyDefinition ReadAssembly (string filePath)
+	{
+		bool haveDebugSymbols = loadDebugSymbols && File.Exists (Path.ChangeExtension (filePath, ".pdb"));
+		var loadReaderParams = new ReaderParameters () {
```

Comment:
There is a `this.readerParameters` field, yet it isn't used here.

This method should probably instead be:

```csharp
AssemblyDefinition ReadAssembly (string filePath, ReaderParameters? parameters)
{
    var loadReaderParams = parameters ?? this.readerParameters;
    // …
}
```

The constructor should likewise be updated so that instead of doing `new ReaderParameters()`, it has this expression.

Reply:
`this.readerParameters` is used to initialize the instance... https://github.com/xamarin/xamarin-android/pull/8164/files/0b63e281fe288aff63d30d5f77a10c2a90a1b183#diff-17f1d51046571abf09ccc3eb26d23c6843e1ac5ad35cf2cb5a825b681ec3f3d2R216